### ISSUE TITLE
Deployment fails on CentOS 7 w/ Qt 5.8 based app

### ISSUE THREAD

**Issue by mhoeher** [probonopd/linuxdeployqt#70]
I tried to bundle a Qt 5.8 based app. Qt has been installed using the upstream installer from qt.io. For deployment, I copied the executable to bundle into a separate directory (`/home/mhoeher/Projects/raspishell/libs/libqvsclient/build/qt5.8-release/deploy`) so I have a clean start. Additionally, I put a *.desktop and *.png file into that directory. Running linuxdeployqt (version 2 from the releases page) I get the following error:

```
"/tmp/.mount_NPzMdN/usr/bin:/home/mhoeher/Qt/5.8/gcc_64/bin/:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/opt/scripts:/home/mhoeher/.local/bin:/home/mhoeher/bin:/opt/scripts"
app-binary: "qvsclient"
ERROR: file copy failed from "/home/mhoeher/Projects/rs/libs/libqvsclient/build/qt5.8-release/deploy/./lib/libgthread-2.0.so.0"
ERROR:  to "./lib//libgthread-2.0.so.0"
ERROR: Error stripping "" : ""
ERROR: Error stripping "" : ""
```

Up to the point of failure the process already created a lib and plugins directory containing the following files:

```
ls lib plugins
lib:
libdbus-1.so.3    libicudata.so.56  libQt5Core.so.5  libQt5Network.so.5  libX11-xcb.so.1   libxcb-dri3.so.0     libxcb-randr.so.0   libxcb-sync.so.1    libXext.so.6    libxshmfence.so.1
libfreetype.so.6  libicui18n.so.56  libQt5DBus.so.5  libQt5Qml.so.5      libXau.so.6       libxcb-glx.so.0      libxcb-render.so.0  libxcb-xfixes.so.0  libXfixes.so.3  libXxf86vm.so.1
libglapi.so.0     libicuuc.so.56    libQt5Gui.so.5   libQt5XcbQpa.so.5   libxcb-dri2.so.0  libxcb-present.so.0  libxcb-shape.so.0   libXdamage.so.1     libXi.so.6

plugins:
platforms
```

I already checked my environment (we are using [environment modules](http://modules.sourceforge.net/) which I first suspected to interfere somehow with linuxdeployqt) to make sure I have a "plain" CentOS environment before starting the process.

Any help on this would be appreciated :)

Btw: I already tried the "continuous" release binary listed in the releases section, however, that one fails to start.

**Comment by mhoeher** [comment]
No, the file is not there. Basically, the directory where it tries to copy *from* is the one where it is supposed to copy *to*. And consequentially it fails. Are there any logs I could provide that would help to better understand the issue?

**Comment by probonopd** [comment]
`linuxdeployqt` uses `ldd` to determine where to copy _from_. What does `ldd` output when you run

```
ldd qvsclient
```

on a pristine copy of `qvsclient`, i.e., one that `linuxdeployqt` never ran on?

**Comment by mhoeher** [comment]
This is the output of ldd:

```
ldd ../app/qvsclient 
        linux-vdso.so.1 =>  (0x00007ffd3eb1a000)
        libQt5Gui.so.5 => /home/mhoeher/Qt/5.8/gcc_64/lib/libQt5Gui.so.5 (0x00007ff39cb52000)
        libQt5Qml.so.5 => /home/mhoeher/Qt/5.8/gcc_64/lib/libQt5Qml.so.5 (0x00007ff39c556000)
        libQt5Network.so.5 => /home/mhoeher/Qt/5.8/gcc_64/lib/libQt5Network.so.5 (0x00007ff39c1d0000)
        libQt5Core.so.5 => /home/mhoeher/Qt/5.8/gcc_64/lib/libQt5Core.so.5 (0x00007ff39bab0000)
        libGL.so.1 => /lib64/libGL.so.1 (0x00007ff39b816000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff39b5fa000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007ff39b2f1000)
        libm.so.6 => /lib64/libm.so.6 (0x00007ff39afee000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007ff39add8000)
        libc.so.6 => /lib64/libc.so.6 (0x00007ff39aa17000)
        libz.so.1 => /lib64/libz.so.1 (0x00007ff39a800000)
        librt.so.1 => /lib64/librt.so.1 (0x00007ff39a5f8000)
        libicui18n.so.56 => /home/mhoeher/Qt/5.8/gcc_64/lib/libicui18n.so.56 (0x00007ff39a15e000)
        libicuuc.so.56 => /home/mhoeher/Qt/5.8/gcc_64/lib/libicuuc.so.56 (0x00007ff399da6000)
        libicudata.so.56 => /home/mhoeher/Qt/5.8/gcc_64/lib/libicudata.so.56 (0x00007ff3983c3000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007ff3981be000)
        libgthread-2.0.so.0 => /lib64/libgthread-2.0.so.0 (0x00007ff397fbc000)
        libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007ff397c85000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff39d2ee000)
        libexpat.so.1 => /lib64/libexpat.so.1 (0x00007ff397a5a000)
        libxcb-dri3.so.0 => /lib64/libxcb-dri3.so.0 (0x00007ff397857000)
        libxcb-present.so.0 => /lib64/libxcb-present.so.0 (0x00007ff397654000)
        libxcb-randr.so.0 => /lib64/libxcb-randr.so.0 (0x00007ff397445000)
        libxcb-xfixes.so.0 => /lib64/libxcb-xfixes.so.0 (0x00007ff39723d000)
        libxcb-render.so.0 => /lib64/libxcb-render.so.0 (0x00007ff397033000)
        libxcb-shape.so.0 => /lib64/libxcb-shape.so.0 (0x00007ff396e2e000)
        libxcb-sync.so.1 => /lib64/libxcb-sync.so.1 (0x00007ff396c27000)
        libxshmfence.so.1 => /lib64/libxshmfence.so.1 (0x00007ff396a24000)
        libglapi.so.0 => /lib64/libglapi.so.0 (0x00007ff3967f4000)
        libselinux.so.1 => /lib64/libselinux.so.1 (0x00007ff3965cd000)
        libXext.so.6 => /lib64/libXext.so.6 (0x00007ff3963bb000)
        libXdamage.so.1 => /lib64/libXdamage.so.1 (0x00007ff3961b7000)
        libXfixes.so.3 => /lib64/libXfixes.so.3 (0x00007ff395fb1000)
        libX11-xcb.so.1 => /lib64/libX11-xcb.so.1 (0x00007ff395daf000)
        libX11.so.6 => /lib64/libX11.so.6 (0x00007ff395a70000)
        libxcb-glx.so.0 => /lib64/libxcb-glx.so.0 (0x00007ff395856000)
        libxcb-dri2.so.0 => /lib64/libxcb-dri2.so.0 (0x00007ff395651000)
        libxcb.so.1 => /lib64/libxcb.so.1 (0x00007ff39542e000)
        libXxf86vm.so.1 => /lib64/libXxf86vm.so.1 (0x00007ff395228000)
        libdrm.so.2 => /lib64/libdrm.so.2 (0x00007ff395019000)
        libpcre.so.1 => /lib64/libpcre.so.1 (0x00007ff394db7000)
        libXau.so.6 => /lib64/libXau.so.6 (0x00007ff394bb2000)
```

However, I now encountered another issue. I had to re-setup my virtual machine which I use for building this evening. (Re-set up in this case means I re-installed the very same set of RPM packages I had previously and restored a backup of my home directory right after). With the VM in that state, I cannot reproduce the original error, but instead get this:

```
linuxdeployqt-2-x86_64.AppImage qvsclient -no-strip -appimage 
"/tmp/.mount_s4o8mE/usr/bin:/home/mhoeher/Qt/5.8/gcc_64/bin/:/usr/lib64/qt-3.3/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin:/opt/scripts:/home/mhoeher/.local/bin:/home/mhoeher/bin:/opt/scripts"
app-binary: "qvsclient"
WARNING: appstreamcli is missing, please install it if you want to use AppStream metadata
Desktop file: /home/mhoeher/Projects/libqvsclient/build/testbuild/dist/default.desktop
Name: qvsclient
Icon: default
Exec: AppRun %F
Comment entry not found in desktop file
Comment: (null)
Type: qvsclient
Categories entry not found in desktop file
Categories: (null)
File used for determining architecture: /home/mhoeher/Projects/libqvsclient/build/testbuild/dist/lib/libQt5Gui.so.5
Arch: 86_64
App name for filename: qvsclient
dest_path: qvsclient-x86_64.AppImage
DESTINATION not specified, so assuming qvsclient-x86_64.AppImage
/home/mhoeher/Projects/libqvsclient/build/testbuild/dist should be packaged as qvsclient-x86_64.AppImage
Creating .DirIcon symlink based on information from desktop file
WARNING: AppStream upstream metadata is missing, please consider creating it
         in usr/share/metainfo/default.appdata.xml
         Please see https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html#sect-Quickstart-DesktopApps
         for more information.
not an application
Failed to generate a template
```

Could the issue be that appstreamcli is missing?

**Comment by probonopd** [comment]
Try running with `-n` to skip the AppStream stuff, please let me know how it goes.

**Comment by mhoeher** [comment]
Sorry for spamming, but: This is an option to one of the underlying tools linuxdeployqt calls? At least the tool itself does not accept it.

**Comment by probonopd** [comment]
Ah sorry, it is an option for `appimagetool` indeed. I need to make it possible to pass this through from `linuxdeployqt` to `appimagetool`.

**Comment by mhoeher** [comment]
Is there a pre-compiled set of the required tools easily available I could test? I currently have only very limited time (we have a deadline approaching end of the month...), but if something is quickly available I could test and report if that helps.

**Comment by probonopd** [comment]
After running the (failed) `linuxdeployqt`, you can run `[appimagetool](https://github.com/probonopd/AppImageKit/releases/download/7/appimagetool-x86_64.AppImage)` on the directory like so:

```
./appimagetool-x86_64.AppImage /home/mhoeher/Projects/libqvsclient/build/testbuild/dist -n
```

Then `-n` makes it skip all things AppStream, I'd be interested if this resolves it for you.

**Comment by mhoeher** [comment]
Good news: Running appimagetool with the `-n` option in fact worked.
Bad news: When I now run linuxdeployqt, I have the original error (failed copying some libraries) back :o I have currently no clue what is different compared to last evening...

**Comment by probonopd** [comment]
@mhoeher please retest with the latest continuous build of `linuxdeployqt`. Does the issue persist?

**Comment by jiakuan** [comment]
I'm getting similar errors with the latest continuous build.

**Operating system**: Ubuntu 16.04 (`Linux ubuntu 4.4.0-72-generic #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`)
**linuxdeployqt version**: Latest linuxdeployqt-continuous-x86_64.AppImage (From [this build](https://travis-ci.org/probonopd/linuxdeployqt/builds/219430489/))

When I tried to create an AppImage using this command:
```
linuxdeployqt-continuous-x86_64.AppImage WiseTime
```

I got this error:

```
Not using FHS-like mode
app-binary: "/home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/WiseTime"
appDirPath: "/home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45"
relativeBinPath: "WiseTime"
ERROR: ldd outputLine: "libmysqlclient.so.18 => not found"
ERROR: Please ensure that all libraries can be found by ldd. Aborting.
```

And here is the `ldd` output:

```
jake@ubuntu:~/work/wisetime-app/dist/WiseTime-2.3.4.45$ ldd WiseTime 
	linux-vdso.so.1 =>  (0x00007ffc76755000)
	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007f4c136bf000)
	libQt5Widgets.so.5 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libQt5Widgets.so.5 (0x00007f4c12e8e000)
	libQt5Gui.so.5 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libQt5Gui.so.5 (0x00007f4c126f2000)
	libQt5Network.so.5 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libQt5Network.so.5 (0x00007f4c1236c000)
	libQt5Sql.so.5 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libQt5Sql.so.5 (0x00007f4c12125000)
	libQt5Core.so.5 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libQt5Core.so.5 (0x00007f4c11a04000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f4c11682000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4c1146c000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4c110a2000)
	libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f4c10e80000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4c10c7c000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4c10a5e000)
	libGL.so.1 => /usr/lib/x86_64-linux-gnu/mesa/libGL.so.1 (0x00007f4c10749000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4c10440000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f4c10225000)
	libicui18n.so.56 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libicui18n.so.56 (0x00007f4c0fd8c000)
	libicuuc.so.56 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libicuuc.so.56 (0x00007f4c0f9d3000)
	libicudata.so.56 => /home/jake/work/wisetime-app/dist/WiseTime-2.3.4.45/./lib/libicudata.so.56 (0x00007f4c0dff0000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f4c0dde8000)
	libgthread-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgthread-2.0.so.0 (0x00007f4c0dbe5000)
	libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f4c0d8d4000)
	/lib64/ld-linux-x86-64.so.2 (0x0000561fb596d000)
	libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007f4c0d6cf000)
	libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f4c0d4c9000)
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f4c0d258000)
```

Looks like it was trying to find libmysqlclient.so.18, but libmysqlclient.so.18 was not shown in `ldd` output at all. Is it possible to exclude `libmysqlclient.so.18` when creating the AppImage? Thanks.

**Comment by jiakuan** [comment]
After I turned on debug, I found it fails when trying to inspecting libqsqlmysql.so

```
Log: Using ldd:
Log:  inspecting "/home/jake/Qt/5.8/gcc_64/plugins/sqldrivers/libqsqlmysql.so"
ERROR: ldd outputLine: "libmysqlclient.so.18 => not found"
ERROR: Please ensure that all libraries can be found by ldd. Aborting.
```

Perhaps it's because my QT project uses `sql` module with SQLite, but no MySQL is required. Is it possible to make it as warning level without exiting directly, or to provide an option to ignore specific dependencies?

**Comment by jiakuan** [comment]
After some research, I've fixed the issue by ignoring unnecessary libraries in QT folder:
```
cd ~/Qt/5.8/gcc_64/plugins/sqldrivers
mv libqsqlmysql.so libqsqlmysql.so_ignore
mv libqsqlpsql.so libqsqlpsql.so_ignore
```

**Comment by mhoeher** [comment]
Hi @probonopd,

sorry for the long time silence, I just didn't find the time to retest until now. Anyway, I gave the new continuous version a try and for the app in question the creation of the app image now works without issues (the tool itself runs through and testing the resulting AppImage on another computer also worked without issues).

**Comment by yooyo3d** [comment]
Main problem is if you have installed QT libs all around system. linuxdeployqt for some reason uses wrong qmake and from there it extract wrong paths. 
I set and export PATH to use my fresh new Qt, and try command qmake -v:
`QMake version 3.1
Using Qt version 5.9.1 in /home/yooyo/Qt/5.9.1/gcc_64/lib`
But when I run linuxdeployqt it uses wrong qmake:
`Log: Using qmake to determine the location of the Qt to be bundled
Log: qmake 5
Log: Using qmake:  "/usr/bin/qmake-qt5"
Log: -query output from qmake: "QT_SYSROOT:\nQT_INSTALL_PREFIX:/usr/lib64/qt5\nQT_INSTALL_ARCHDATA:/usr/lib64/qt5\nQT_INSTALL_DATA:/usr/share/qt5\nQT_INSTALL_DOCS:/usr/share/doc/qt5\nQT_INSTALL_HEADERS:/usr/include/qt5\nQT_INSTALL_LIBS:/usr/lib64\nQT_INSTALL_LIBEXECS:/usr/lib64/qt5/libexec\nQT_INSTALL_BINS:/usr/lib64/qt5/bin\nQT_INSTALL_TESTS:/usr/lib64/qt5/tests\nQT_INSTALL_PLUGINS:/usr/lib64/qt5/plugins\nQT_INSTALL_IMPORTS:/usr/lib64/qt5/imports\nQT_INSTALL_QML:/usr/lib64/qt5/qml\nQT_INSTALL_TRANSLATIONS:/usr/share/qt5/translations\nQT_INSTALL_CONFIGURATION:/etc/xdg\nQT_INSTALL_EXAMPLES:/usr/lib64/qt5/examples\nQT_INSTALL_DEMOS:/usr/lib64/qt5/examples\nQT_HOST_PREFIX:/usr/lib64/qt5\nQT_HOST_DATA:/usr/lib64/qt5\nQT_HOST_BINS:/usr/lib64/qt5/bin\nQT_HOST_LIBS:/usr/lib64\nQMAKE_SPEC:linux-g++\nQMAKE_XSPEC:linux-g++\nQMAKE_VERSION:3.0\nQT_VERSION:5.6.1\n"
Log: Qt libs path determined from qmake: "/usr/lib64"
Log: Changed LD_LIBRARY_PATH: "/usr/lib64:"`

Then it copy wrong libs to lib folder and resulting app-image not working (complaints about missing libQt5Core.so.5)

**Comment by mhoeher** [comment]
Hi @yooyo3d,

this could be due to the heuristic the tool uses to find the right `qmake` executable. The system default `qmake` is often called `qmake-qt5` or `qmake-qt4` on some distributions. IIRC, the code for finding `qmake` first searches for the variants with suffix and only if that fails, it looks for an executable named `qmake`.

Long story short: In your case try to use the `-qmake` command line switch to explicitly point `linuxdeployqt` to the location where your Qt is installed:

```bash
linuxdeployqt -qmake /home/yooyo/Qt/5.9.1/gcc_64/bin/qmake # other command line args....
```
